internal/security/secretstore/tokenfilewriter/tokenfilewriter.go

```diff
@@ -0,0 +1,172 @@
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/clients/logger"
+	"github.com/edgexfoundry/go-mod-secrets/v2/pkg/token/fileioperformer"
+	"github.com/edgexfoundry/go-mod-secrets/v2/secrets"
+)
+
+const (
+	consulSecretsEngineOpsPolicyName = "consul_secrets_engine_management_policy"
+)
+
+// TokenFileWriter is a mechanism to generates a token and writes it into a file specified by configuration
+type TokenFileWriter struct {
+	logClient    logger.LoggingClient
+	secretClient secrets.SecretStoreClient
+	fileOpener   fileioperformer.FileIoPerformer
+}
+
+// NewWriter instantiates a TokenFileWriter instance
+func NewWriter(lc logger.LoggingClient,
+	sc secrets.SecretStoreClient,
+	fileOpener fileioperformer.FileIoPerformer) TokenFileWriter {
+	return TokenFileWriter{
+		logClient:    lc,
+		secretClient: sc,
+		fileOpener:   fileOpener,
+	}
+}
+
+// CreateAndWrite generates a new token and writes it to the file specified by tokenFilePath
+// the generation of the token requires root token privilege
+// it overwrites the file if already exists
+// returns error if anything fails during the whole process
+func (w TokenFileWriter) CreateAndWrite(rootToken string, tokenFilePath string,
+	createTokenFunc tokencreatable.CreateTokenFunc) (tokencreatable.RevokeFunc, error) {
+	if len(rootToken) == 0 {
+		return nil, fmt.Errorf("rootToken is required")
+	}
+
+	// creates the token
+	createTokenFuncName := getFunctionName(createTokenFunc)
+	tokenCreated, revokeTokenFunc, createErr := createTokenFunc(rootToken)
+	if createErr != nil {
+		return nil, fmt.Errorf("failed to create token with %s: %s", createTokenFuncName, createErr.Error())
+	}
+
+	w.logClient.Infof("created token with %s", createTokenFuncName)
+
+	var fileErr error
+	defer func() {
+		// call revokeTokenFunc if there is some fileErr and revokeTokenFunc itself is not nil
+		if fileErr != nil && revokeTokenFunc != nil {
+			revokeTokenFunc()
+		}
+	}()
+
+	// Write the token issuing token to disk to pass it to the token provider
+	tokenFileAbsPath, fileErr := filepath.Abs(tokenFilePath)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to convert tokenFile to absolute path %s: %s", tokenFilePath, fileErr.Error())
+	}
+
+	dirOfCreatedToken := filepath.Dir(tokenFileAbsPath)
+	fileErr = w.fileOpener.MkdirAll(dirOfCreatedToken, 0700)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create tokenpath base dir: %s", fileErr.Error())
+	}
+
+	fileWriter, fileErr := w.fileOpener.OpenFileWriter(tokenFileAbsPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0600)
+	if fileErr != nil {
+		return nil, fmt.Errorf("failed to create token file %s: %s", tokenFileAbsPath, fileErr.Error())
+	}
+
+	if fileErr = json.NewEncoder(fileWriter).Encode(tokenCreated); fileErr != nil {
+		_ = fileWriter.Close()
+		return nil, fmt.Errorf("failed to write created token: %s", fileErr.Error())
+	}
+
+	if fileErr = fileWriter.Close(); fileErr != nil {
```

Comment:
See internal/security/fileprovider/provider.go lines 172-196.  Since we are passing the token to another container, we want to be able to configure file permissions, uid, and gid.

Reply:
I don't understand this: the existing code for creating token-issuing token doesn't have this logic at all;  are you suggesting we modify the existing privilege for token issuing token , too???

Reply:
@bnevis-i  need more discussion on this. I do not quite understand the logic in the line 172-196 you mentioned above there and I thought the file permission already specified by the fileOpenWriter.  We are not using this token to create another token so I don't understand why we need to do that step. Let's have a discussion tomorrow morning if you have time.

Reply:
fileOpenWriter only specifies the mode.  If we are running containers as non-root users the file owner and group must match that of the consuming container.

Reply:
Hi @bnevis-i  if the `secretstore-setup` is running as non-root, then the chown operations won't work, right?  Also as of today, this code in the entrypoint script running as root, https://github.com/edgexfoundry/edgex-go/blob/master/cmd/security-secretstore-setup/entrypoint.sh#L33-L36  would change the ownership to 2002:2001 anyway, so that the ownership change in the provider is a moot point.

Reply:
resolve as we discussed